feat(bitbucket-server): Support gitUrl

jatladams opened the pull request:
## Changes:

This allows the gitUrl property to be used in bitbucket-server to set how repositories should be accessed.

If no value is provided for gitUrl the previous logic will be used. This should mean users who upgrade without change see no difference in functionality.

If a gitUrl value is provided as per https://docs.renovatebot.com/self-hosted-configuration/#giturl:
default: use HTTPS URLs provided by the platform for Git or return undefined if none provided
ssh: use SSH URLs provided by the platform for Git or return undefined if none provided
endpoint: ignore URLs provided by the platform and use the configured endpoint directly

Added tests to 'utils.spec.ts' to cover 'getRepoGitUrl'
Added tests to 'index.spec.ts' to cover 'gitUrl' values

## Context:

Currently there is no way to specify a specific git connection type. Because Bitbucket Server provides all available connection types in the [API response ](https://github.com/renovatebot/renovate/blob/9e804e9cb7e9ed12fb66f031dd4e611cc8726efb/lib/modules/platform/bitbucket-server/index.ts#L179)  (http and ssh) the only way to force ssh connection currently is to turn off http connection.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

<!-- Do you have any suggestions about this PR template? Edit it here: https://github.com/renovatebot/renovate/edit/main/.github/pull_request_template.md -->

<!-- Please do not force push to your PR's branch after you have created your PR, as doing so forces us to review the whole PR again. This makes it harder for us to review your work because we don't know what has changed. -->
<!-- PRs will always be squashed by us when we merge your work. Commit as many times as you need in this branch. -->

jatladams committed:
feat: Support gitUrl on bitbucket-server platform

This allows the gitUrl property to be used in bbs to select how repositories should be accessed

rarkins reviewed (changes requested):
Needs docs added, but first please create an issue to discuss your requirements and proposed implementation approach

jatladams commented:
@rarkins There is an existing issue for this change: https://github.com/renovatebot/renovate/issues/9607
Is there a way to attach this change to that issue?

rarkins commented:
You can type `Closes #` in the Context of PRs to link the two. However that issue has label `status:requirements` which means it needs further discussion before confirming it's ready for implementation

jatladams commented:
Ah I missed that tag, my mistake.

I saw the issue and just followed the implementation in `https://github.com/renovatebot/renovate/blob/main/lib/platform/gitlab/index.ts getRepoUrl()`

Are there any steps I can take to move the issue forward?

viceice commented:
Please do the refactoring first in a separate PR, so we can more easily review.

jatladams commented:
Is this what you had in mind @viceice ?
https://github.com/renovatebot/renovate/commit/abbe1773b59a4b8a075bac051307780a05b6af2b?diff=split

viceice commented:
Yes, the getRepoUrl function extraction

harmjanblok commented:
Seems already good progress has been made on this issue. What are the remaining open points for this to be merged?

viceice reviewed (changes requested):
needs deconflicting

👉 https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review

viceice commented on lib/platform/bitbucket-server/index.ts:
```diff
@@ -153,11 +159,60 @@ export async function getJsonFile(
   return JSON.parse(raw);
 }
 
+function getRepoUrl(
```
Please do this function extraction first in a separate PR without any functional change. Then use this branch / PR for the  gitUrl` changes only.

rarkins committed:
Merge branch 'main' into bbs-ssh

# Conflicts:
#	lib/modules/platform/bitbucket-server/index.ts

jatladams commented:
Apologies I haven't had a chance to fix up the [refactor](https://github.com/jatladams/renovate/tree/bbs-index-refactor) requested to move this forward. I will try find some time this week to work on it

rarkins commented:
Needs deconflicting

jatladams committed:
Merge branch 'main' into bbs-ssh

# Conflicts:
#	lib/modules/platform/bitbucket-server/index.ts

jatladams committed:
Merge branch 'renovatebot:main' into bbs-ssh

jatladams commented:
Hi @rarkins and @viceice,
I have resolved the merge conflicts, updated the logic so that users will so no change unless they specify a value for gitUrl and added unit tests to ensure coverage of the new code.

For the new functionality I have matched the logic as specified in https://docs.renovatebot.com/self-hosted-configuration/#giturl or return undefined if the expected values are not returned from Bitbucket Server API

jatladams commented on lib/modules/platform/bitbucket-server/index.spec.ts:
```diff
@@ -291,6 +291,117 @@ describe('modules/platform/bitbucket-server/index', () => {
           expect(httpMock.getTrace()).toMatchSnapshot();
         });
 
+        it('no git url', async () => {
```
I think this test may not have been correctly resolved in the merge, I will investigate

jatladams committed:
Merge branch 'main' into bbs-ssh

# Conflicts:
#	lib/modules/platform/bitbucket-server/__snapshots__/index.spec.ts.snap

jatladams committed:
Merge branch 'main' into bbs-ssh

rarkins commented on lib/modules/platform/bitbucket-server/index.ts:
```diff
@@ -191,16 +192,17 @@ export async function initRepo({
     await git.initRepo({
       ...config,
-      url: gitUrl,
+      url: url,
```
```suggestion
      url,
```

rarkins committed:
Add supportedPlatforms

rarkins reviewed (changes requested):
Please also update the description inside docs/usage/

viceice commented on lib/modules/platform/bitbucket-server/index.spec.ts:
```diff
@@ -291,6 +291,117 @@ describe('modules/platform/bitbucket-server/index', () => {
+              repository: 'SOME/repo',
+            })
+          ).toMatchSnapshot();
+          expect(httpMock.getTrace()).toMatchSnapshot();
```
remove

viceice commented on lib/modules/platform/bitbucket-server/index.spec.ts:
```diff
@@ -291,6 +291,117 @@ describe('modules/platform/bitbucket-server/index', () => {
+              endpoint: 'https://stash.renovatebot.com/vcs/',
+              repository: 'SOME/repo',
+            })
+          ).toMatchSnapshot();
```
use `toEqual` matcher

viceice commented on lib/modules/platform/bitbucket-server/index.spec.ts:
```diff
@@ -291,6 +291,117 @@ describe('modules/platform/bitbucket-server/index', () => {
+          expect(git.initRepo).toHaveBeenCalledWith(
+            expect.objectContaining({ url: sshLink('SOME', 'repo') })
+          );
+          expect(res).toMatchSnapshot();
+          expect(httpMock.getTrace()).toMatchSnapshot();
```
same here. please also fix all other new/ changed tests.

viceice commented on lib/modules/platform/bitbucket-server/utils.spec.ts:
```diff
@@ -1,11 +1,70 @@
+            name: 'ssh',
+          }
+        : null,
+    ].filter(Boolean);
```
use `is.truthy`

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -155,38 +157,80 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
+): string {
+  const { host, pathname } = url.parse(defaultEndpoint);
+  const generatedUrl = git.getUrl({
+    protocol: defaultEndpoint.split(':')[0] as GitProtocol,
```
why not using from parsed url?

jatladams replied:
Good point, I also changed the deprecated `url.parse(defaultEndpoint);` for `new URL(defaultEndpoint);`

viceice replied:
use `parseUrl` from `util/url`

viceice replied:
and please do the refactoring / preparation in a separate PR to be merged first

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -155,38 +157,80 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
+      // Fallback to generating the url if the API didn't give us an URL
+      return generateUrlFromEndpoint(defaultEndpoint, opts, repository);
+    }
+    return undefined;
```
use `null` instead of `undefined`

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -155,38 +157,80 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
+  gitUrl: GitUrlOption,
   info: BbsRestRepo,
   opts: HostRule
 ): string {
```
```suggestion
 ): string | null {
```

jatladams committed:
Apply PR feedback
- Remove httpMock expect from tests
- Replace toMatchSnapshot with toEqual matcher
- Replace deprecated url.parse with new URL()
- Return null from getRepoGitUrl() instead of undefined

jatladams committed:
Merge branch 'renovatebot:main' into bbs-ssh

jatladams commented:
I have applied the feedback, I still need to update the docs
- [ ] update docs/usage/

jatladams committed:
Cleaned up index.spec.ts.snap

jatladams committed:
Merge branch 'bbs-ssh' of github.com:jatladams/renovate into bbs-ssh

rarkins commented:
Needs deconflicting

jatladams committed:
Merge branch 'main' into bbs-ssh

jatladams committed:
- Update docs to show bitbucket server supports gitUrl
- add exception thrown if git url not found in getRepoGitlUrl

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -155,39 +157,81 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
 export function getRepoGitUrl(
   repository: string,
   defaultEndpoint: string,
+  gitUrl: GitUrlOption | undefined,
```
this function can be simplified, use a switch case on `gitUrl` and throw a config error if `gitUrl` is set, but type couldn't found ( http, ssh )

jatladams replied:
Do you mean something like:
`  switch(gitUrl) {
    case "default": {
      // do something;
      break;
    }
    case "ssh": {
      // do something;
      break;
    }
    case "endpoint": {
      // do something;
      break;
    }
    case undefined: {
      // do something;
      break;
    }
    default: {
      //throw error;
      break;
    }
  }`

jatladams replied:
Changes applied. 

I used undefined to follow the existing logic where we try http then ssh and finally generate the URL.

rarkins commented on docs/usage/self-hosted-configuration.md:
```diff
@@ -370,7 +370,7 @@ To handle the case where the underlying Git processes appear to hang, configure
 
 Override the default resolution for Git remote, e.g. to switch GitLab from HTTPS to SSH-based.
-Currently works for GitLab only.
+Currently works for Bitbucket Server and GitLab only.
```
```suggestion
```

Add `supportedPlatform` to its config options definition instead

jatladams replied:
Is there another config option to update besides https://github.com/renovatebot/renovate/pull/13875/commits/affda143098433a19f6d249d01ac0cf21528fd3c ?

rarkins replied:
No other

jatladams committed:
Merge branch 'renovatebot:main' into bbs-ssh

mintmac commented:
I'm also after this feature. Any chances to push this forward?

viceice commented:
needs deconfliction

submac commented:
I have a very little knowledge of js/ts. Could @jatladams or someone else help to close it?

viceice committed:
Merge branch 'main' into bbs-ssh

rarkins committed:
Merge branch 'main' into bbs-ssh

rarkins commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -156,39 +158,81 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
-    cloneUrl = info.links.clone?.find(({ name }) => name === 'ssh');
-  }
+): string | null {
+  if (gitUrl === undefined) {
```
@viceice do the latest typing changes mean that this is now not possible - would be `null`?

jatladams committed:
Merge branch 'renovatebot:main' into bbs-ssh

jatladams commented:
@submac I can look at the `utils.ts` that are requested and the lint failures.

jatladams committed:
Update getRepoGitUrl to use switch statement
Updated infoMock to fix type incompatibility issues

jatladams committed:
Merge branch 'main' into bbs-ssh

jatladams committed:
Merge branch 'main' into bbs-ssh

jatladams committed:
Add test to cover default use case in getRepoGitUrl
Remove Unreachable code

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
+      const sshUrl = info.links.clone?.find(({ name }) => name === 'ssh');
+      if (sshUrl === undefined) {
+        return null;
+      }
+
+      logger.debug({ url: sshUrl.href }, `using ssh URL`);
+      return sshUrl.href;
+    }
+    case 'endpoint': {
+      const generatedUrl = generateUrlFromEndpoint(
+        defaultEndpoint,
+        opts,
+        repository
+      );
+      logger.debug({ url: generatedUrl }, `using endpoint URL`);
+      return generatedUrl;
+    }
+    case undefined: {
+      //if no connection type is specified try them all as a fallback
+      const httpUrl = getRepoGitUrl(
+        repository,
+        defaultEndpoint,
+        'default',
+        info,
+        opts
+      );
+      if (httpUrl !== null) {
+        return httpUrl;
+      }
+
+      const sshUrl = getRepoGitUrl(
+        repository,
+        defaultEndpoint,
+        'ssh',
+        info,
+        opts
+      );
+      if (sshUrl !== null) {
+        return sshUrl;
+      }
```
```suggestion
): string {
  if (gitUrl === 'ssh') {
    const sshUrl = info.links.clone?.find(({ name }) => name === 'ssh');
    if (sshUrl === undefined) {
      throw new Error(CONFIG_GIT_URL_UNAVAILABLE);
    }

    logger.debug({ url: sshUrl.href }, `using ssh URL`);
    return sshUrl.href;
  }

  let cloneUrl = info.links.clone?.find(({ name }) => name === 'http');
  if (cloneUrl) {
    // Inject auth into the API provided URL
    return injectAuth(cloneUrl.href, opts);
  }

  // Http access might be disabled, try to find ssh url in this case
  cloneUrl = info.links.clone?.find(({ name }) => name === 'ssh');

  if (gitUrl === 'endpoint' || !cloneUrl) {
    return generateUrlFromEndpoint(defaultEndpoint, opts, repository);
  }

  // SSH urls can be used directly
  return cloneUrl.href;
}
```

jatladams replied:
I am slightly concerned a user could set `gitUrl` to `endpoint` and it would be ignored if `http` or `ssh` are available. Is this the desired functionality?

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
+  });
+  logger.debug({ url: generatedUrl }, `using generated endpoint URL`);
+  return generatedUrl;
+}
```
```suggestion
}

function injectAuth(url: string, opts: HostRule): string {
  const repoUrl = parseUrl(url)!;
  if (!repoUrl) {
    logger.debug(`Invalid url: ${url}`);
    throw new Error(CONFIG_GIT_URL_UNAVAILABLE);
  }
  // TODO: null checks (#7154)
  repoUrl.username = opts.username!;
  repoUrl.password = opts.password!;
  return repoUrl.toString();
}
```

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -11,6 +12,7 @@ import type {
   HttpResponse,
 } from '../../../util/http/types';
 import { getPrBodyStruct } from '../pr-body';
```
```suggestion
import { ensureTrailingSlash, parseUrl } from '../../../util/url';
import { getPrBodyStruct } from '../pr-body';
```

jatladams replied:
`ensureTrailingSlash` doesn't appear to be used, is it missing from another change?

viceice replied:
yes, sorry, leftover from testing

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
-  } else {
-    // SSH urls can be used directly
-    gitUrl = cloneUrl.href;
+      return getRepoGitUrl(repository, defaultEndpoint, 'endpoint', info, opts);
+    }
+    default: {
+      throw new TypeError(
+        `The specified scm connection type ${gitUrl} is not a valid GitUrlOption value`
+      );
+    }
   }
```
```suggestion
```

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -156,39 +158,93 @@ export function getInvalidReviewers(err: BitbucketError): string[] {
+    }
   }
-  return gitUrl;
 }
```
```suggestion
```

viceice reviewed (commented):
simplify things, sorry switch made it even more complex

viceice commented on lib/modules/platform/bitbucket-server/index.ts:
```diff
@@ -189,17 +191,22 @@ export async function initRepo({
       opts
     );
 
+    if (url === null) {
+      throw new Error(CONFIG_GIT_URL_UNAVAILABLE);
+    }
+
```
```suggestion
```

viceice commented on lib/modules/platform/bitbucket-server/index.ts:
```diff
@@ -4,6 +4,7 @@ import JSON5 from 'json5';
 import type { PartialDeep } from 'type-fest';
 import { PlatformId } from '../../../constants';
 import {
+  CONFIG_GIT_URL_UNAVAILABLE,
```
```suggestion
```

viceice commented on lib/modules/platform/bitbucket-server/utils.ts:
```diff
@@ -1,6 +1,7 @@
 // SEE for the reference https://github.com/renovatebot/renovate/blob/c3e9e572b225085448d94aa121c7ec81c14d3955/lib/platform/bitbucket/utils.js
 import url from 'url';
 import is from '@sindresorhus/is';
```
```suggestion
import is from '@sindresorhus/is';
import { CONFIG_GIT_URL_UNAVAILABLE } from '../../../constants/error-messages';
```

jatladams committed:
Merge branch 'main' into bbs-ssh

jatladams committed:
Merge branch 'main' into bbs-ssh

jatladams committed:
Apply PR feedback to simplify getRepoGitUrl and move error handling inside utils.ts for CONFIG_GIT_URL_UNAVAILABLE

viceice committed:
Merge branch 'main' into bbs-ssh

rarkins committed:
Merge branch 'main' into bbs-ssh

rarkins committed:
Merge branch 'main' into bbs-ssh
